EventGenerator/src/RPCGun_module.cc

```diff
@@ -56,214 +54,208 @@ namespace mu2e {
+          _hy        = tfdir.make<TH1F>("hy"       , "y = (ee-ep)/|pe+pp|", 200,-1.,1.);      
         }
       }
-
-    }
-
-  //================================================================
-
-  void RPCGun::beginRun(art::Run& run) {
-    protonPulse_.reset( new ProtonPulseRandPDF( eng_, protonPset_ ) );
   }
-
-  void RPCGun::produce(art::Event& event) {
-
-    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
-    std::unique_ptr<FixedTimeMap> timemap(new FixedTimeMap);
-
-    ConditionsHandle<AcceleratorParams> accPar("ignored");
-    double _mbtime = accPar->deBuncherPeriod;
-
-    IO::StoppedParticleTauNormF stop;
-    if (tmin_ > 0){
-      while (true){
-        const auto& tstop = stops_.fire();
-        timemap->SetTime(protonPulse_->fire());
-        if (tstop.t+timemap->time() < 0 || tstop.t+timemap->time() > tmin_){
-          if (applySurvivalProbability_){
-            double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-            if (weight > 1)
-              std::cout << "WEIGHT TOO HIGH " << weight << " " << fmod(tstop.t,_mbtime) << std::endl;
-            double rand = randomFlat_.fire();
-            if (weight > rand){
-              stop = tstop;
-              break;
+  
+  double RPCGun::MakeEventWeight(art::Event& event){ 
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    double weight = 0.;
+    double tau = 0.;
+    const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+    for(const auto& p : *simh) {
+      if(p.second.daughters().empty()) {
+          art::Ptr<SimParticle> part(simh, p.first.asUint());
+          std::cout<<"part info: "<<part->creationCode()<<" "<<part->pdgId()<<std::endl;
+          tau = part->endProperTime() / gc.getParticleLifetime(part->pdgId());
+          while(part->parent().isNonnull()) {
+            if((part->creationCode() == ProcessCode::mu2ePrimary)) {
+              part = part->parent();
+              tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId());
             }
-          }else{
-            stop = tstop;
-            break;
+            else {
+              part = part->parent();
+              if ( std::binary_search(decayOffPDGCodes_.begin(), decayOffPDGCodes_.end(), int(part->pdgId()) ) ) {
+                tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId()); 
+              }
           }
-        }
-      }
-    }else{
-      timemap->SetTime(protonPulse_->fire());
-      if (applySurvivalProbability_){
-        while (true){
-          const auto& tstop = stops_.fire();
-          double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-          double rand = randomFlat_.fire();
-          if (weight > rand){
-            stop = tstop;
-            break;
-          }
-        }
-      }else{
-        const auto& tstop = stops_.fire();
-        stop = tstop;
+        } 
+        weight += exp(-tau);
```

Comment:
Why is the weight the sum of the individual sim particle weights? Should it be the product of them? Is there any risk of double counting weights due to sim particle sharing some parent?